llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp

```diff
@@ -937,6 +951,43 @@ std::optional<SIMemOpInfo> SIMemOpAccess::getAtomicCmpxchgOrRmwInfo(
   return constructFromMIWithMMO(MI);
 }
 
+std::optional<std::tuple<SIMemOpInfo, SIMemOp>>
+SIMemOpAccess::getLdsLoadStoreInfo(
+    const MachineBasicBlock::iterator &MI) const {
+  assert(MI->getDesc().TSFlags & SIInstrFlags::maybeAtomic);
+
+  if (!(MI->mayLoad() && MI->mayStore()))
+    return std::nullopt;
+
+  // An LDS DMA will have exactly two memory operands.
+  if (MI->getNumMemOperands() != 2)
+    return std::nullopt;
```

Comment:
Better to make the machine verifier enforce 0 or 2

Reply:
ig isLDSDMA()? https://github.com/llvm/llvm-project/blob/c50ed05cada308dea282e026cc782a9ea3ab0c29/llvm/lib/Target/AMDGPU/SIInstrInfo.h#L594

Reply:
... Ah. I'll put the asserts back then